advanced_source/torch_script_custom_classes.rst

```diff
@@ -0,0 +1,399 @@
+demonstrates that:
+
+.. code-block:: python
+
+  import torch
+
+  # `torch.classes.load_library()` allows you to pass the path to your .so file
+  # to load it in and make the custom C++ classes available to both Python and
+  # TorchScript
+  torch.classes.load_library("libcustom_class.so")
+
+  # We can find and instantiate our custom C++ class in python by using the
+  # `torch.classes` namespace:
+  #
+  # This instantiation will invoke the Stack(std::vector<T> init) constructor
+  # we registered earlier
+  s = torch.classes.Stack(["foo", "bar"])
+
+  # We can call methods in Python
+  s.push("pushed")
+  assert s.pop() == "pushed"
+
+  # Returning and passing instances of custom classes works as you'd expect
+
+  s2 = s.clone()
+  s.merge(s2)
+  for expected in ["bar", "foo", "bar", "foo"]:
+      assert s.pop() == expected
+
+  # We can also use the class in TorchScript
+  # For now, we need to assign the class's type to the local in order to
+  # annotate the type on the TorchScript function
+  Stack = torch.classes.Stack
+
+  # This demonstrates:
+  #   - passing a custom class instance to TorchScript
+  #   - instantiating a class in TorchScript
+  #   - calling a custom class's methods in torchscript
+  #   - returning a custom class instance from TorchScript
+  @torch.jit.script
+  def do_stacks(s : Stack):
+      s2 = torch.classes.Stack(["hi", "mom"])
+      s2.merge(s)
+      return s2.clone(), s2.top()
+
+  stack, top = do_stacks(torch.classes.Stack(["wow"]))
+  assert top == "wow"
+  for expected in ["wow", "mom", "hi"]:
+      assert stack.pop() == expected
+
+Saving, Loading, and Running TorchScript Code Using Custom Classes
+------------------------------------------------------------------
+
+We can also use custom-registered C++ classes in a C++ process using
+libtorch. As an example, let's define a simple ``nn.Module`` that
+instantiates and calls a method on our Stack class:
+
+.. code-block:: python
+
+  import torch
+
+  torch.classes.load_library('libcustom_class.so')
+
+  class Foo(torch.nn.Module):
+      def __init__(self):
+          super().__init__()
+
+      def forward(self, s : str) -> str:
+          stack = torch.classes.Stack(["hi", "mom"])
+          return stack.pop() + s
+
+  scripted_foo = torch.jit.script(Foo())
+  print(scripted_foo.graph)
+
+  scripted_foo.save('foo.pt')
+
+``foo.pt`` in our filesystem now contains the serialized TorchScript
+program we've just defined.
+
+Now, we're going to define a new CMake project to show how you can load
+this model and its required .so file. For a full treatment of how to do this,
+please have a look at the `Loading a TorchScript Model in C++ Tutorial <https://pytorch.org/tutorials/advanced/cpp_export.html>`_.
+
+Similarly to before, let's create a file structure containing the following::
+
+  cpp_inference_example/
+    infer.cpp
+    CMakeLists.txt
+    foo.pt
+    build/
+    custom_class_project/
```

Comment:
Can you fully specify what's in `custom_class_project/` just so everything is explicit?